solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -375,10 +374,152 @@ private StatefulCollectionWatch compute(
+          log.warn("ZooKeeper watch triggered, but Solr cannot talk to ZK: ", e);
+        } catch (KeeperException e) {
+          log.error("exception for collection: [{}]", coll, e);
+          throw new ZooKeeperException(ErrorCode.SERVER_ERROR, "A ZK error has occurred", e);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          log.error("exception for collection: [{}]", coll, e);
+        }
+        collectionWatches.updateDocCollection(coll, newState);
+        synchronized (getUpdateLock()) {
+          constructState(Collections.singleton(coll));
+        }
+      }
+      if (collectionState != null && collectionState.isPerReplicaState()) {
+        String path = event.getPath();
+        if ((event.getType() == EventType.NodeCreated || event.getType() == EventType.NodeDeleted)
+            && path.length() > collectionState.getZNode().length()
+            && path.startsWith(collectionPath)) {
+          if (log.isDebugEnabled()) {
+            log.debug("PRS node event : {}", event.getType());
+          }
+
+          PerReplicaStates prs = collectionState.getPerReplicaStates();
+          if (prs != null) {
+            String stateStr = path.substring(collectionPath.length() + 1);
+            PerReplicaStates.State newState = PerReplicaStates.State.parse(stateStr);
+            PerReplicaStates.State oldState = prs.states.get(newState.replica);
+            if (event.getType() == EventType.NodeCreated) {
+              if (oldState != null && newState.version < oldState.version) {
+                // we got a notification out of order? . Shouldn't happen
+                if (log.isInfoEnabled()) {
+                  log.info("newState {} < oldState {}", newState, oldState);
+                }
+                return;
+              }
+              if (oldState == null) {
+                // the state does not exist now. fetch everything
+
+                if (log.isTraceEnabled()) {
+                  log.trace("fresh replica, force fetch all {}", collectionPath);
+                }
+
+                prs = PerReplicaStatesFetcher.fetch(collectionPath, zkClient, null);
+              } else {
+                // the PRS entry is already available, need to do an update
+                Stat stat = null;
+                try {
+                  // get the stat of this child node
+                  stat = zkClient.exists(path, null, true);
+                } catch (Exception e) {
+                  throw new RuntimeException(e);
+                }
+                if (stat == null) {
+                  // this got deleted so soon. nothing to do
+                  return;
+                }
+
+                if (log.isTraceEnabled()) {
+                  log.trace("PRS insert {}, v:{}", newState, stat.getCzxid());
+                }
+                prs = prs.insert(newState, stat.getCzxid());
+                if (prs == null) {
+                  // something went wrong
```

Comment:
This seems to get triggered for every core during collection creation. as preceding create event on the collection invokes line 407 an fetch the collection state with the PerReplicateStates already. Therefore all following PRS path create event will actually find identical entries and failed (https://github.com/cowpaths/fullstory-solr/pull/105/files#diff-0bd8a828302915c525c8df3e8cccdc9881ebad121359c0dbc8374b8b72995669R151, which both version numbers are the same)